Make sensor (get single value) operation return unquoted strings.

geomacy opened the pull request:
Before:

```
 br app salt-test ent myent sensor host.address
"10.10.10.103"
```

After:

```
 br app salt-test ent myent sensor host.address
10.10.10.103
```

geomacy committed:
Make sensor (get single value) operation return unquoted strings.

geomacy commented:
Hi @robertgmoss would appreciate it if you can cast an eye over this and merge if possible.

robertgmoss commented on api/entity_sensors/sensors.go:
```diff
@@ -11,7 +11,9 @@ import (
-	body, err := network.SendGetRequest(url)
+	headers := make(map[string]string)
+	headers["Accept"] = "text/plain"
+	body, err := network.SendGetRequestWithHeaders(url, headers)
```
For the sake of consistency, I'd use `json.Unmarshal` instead of sending different headers, see e.g. `SensorList`.

robertgmoss commented:
@geomacy I'm thinking that we always ask for json from the server and unmarshal it, that way we can add an option not to unmarshal it easily for a json view.  WDYT?

ahgittin commented:
my mistake - this was my suggestions

BTW i think the main differences are that:
* plain returns strings unquoted, and a 0-length string for null values
* json returns strings quoted, and null for null values

Otherwise the return values are pretty much identical I believe.  So JSON unmarshalled is probably better, then decide whether to quote strings or not (and default is not to do so for easier use when scripting?).

lloyddave commented:
FWIW, I would go with unmarshalling (sp?) the JSON and removing the quote marks.

geomacy committed:
Review comment updates, use Json unmarshalling rather than headers to get the desired behaviour.

Also adds the same logic for the 'ConfigValue' operation for convenience.

geomacy commented:
hi @lloyddave, @robertgmoss, @ahgittin, thanks for the comments, I have updated as suggested. 

Also added the same logic for the 'config' operation:

Before:
```
$ br app salt-test ent myent config brooklyn.salt.start.states.pTPvyiPn
"apache"
```

After:
```
$ br app salt-test ent myent config brooklyn.salt.start.states.pTPvyiPn
apache
```

robertgmoss commented on api/entity_config/config.go:
```diff
@@ -12,6 +12,12 @@ func ConfigValue(network *net.Network, application, entity, config string) (stri
+
+	var value string;
+	err = json.Unmarshal(bytes, &value)
+	if nil == err {
```
I personally prefer this to be the `nil != err` case, so that the last thing returned is the happy path, and maybe return the error too.

geomacy replied:
Really both line 19 and 21 are happy paths - 19 is when you can serialise the JSON config value to a string, ("apache"), and 21 when it's more complex, in which case we just return the JSON text. 

I could do it as 

``` go
    if nil != err {
        return string(bytes), nil
    }
   return value, nil
```

if you think that is better style?  Or maybe better with an explicit 'else'?

robertgmoss replied:
I'm not convinced line 21 is a happy path - you tried to unmarshal a string, but got something else unexpected.
I was thinking

```
    if nil != err {
        return string(bytes), err
    }
    return value, nil
```

i.e return the `err` too because calling code might want to know why the unmarshal failed.

geomacy replied:
But then the calling code will see it as a failure - 

``` go
config, err := entity_config.ConfigValue(cmd.network, scope.Application, scope.Entity, c.Args().First())
        if nil != err {
            fmt.Fprintln(os.Stderr, err)
            os.Exit(1)
        }
```

whereas the intent of the code in ConfigValue is to:
1. Try interpreting it as a single JSON string (line 17); if this works, return the string value (line 19).
2. If 1. doesn't work, it is some more complex JSON expression, so just return the literal JSON (line 21).

A failure to unmarshal at line 17 is fine, it just means the value isn't a simple "json string".

robertgmoss replied:
Then I would suggest unmarshalling to type `interface{}` rather than `string` and let the client code examine it with a switch on its type. (cf. http://blog.golang.org/json-and-go)

geomacy committed:
Further review comment updates.

geomacy commented:
That is indeed nicer; I have made the update and pushed it.  How does that look? Tested behaviour is as desired.

geomacy committed:
Further review comment updates.

Update for:

https://github.com/brooklyncentral/brooklyn-cli/commit/9e246f5c10cb31bbcd325a02f3e8dad0006a9bc7#commitcomment-15919777

geomacy commented:
Actually this isn't so easy, note that the values returned in the limbs of the switch are different variables:
```go
switch value.(type) {
	case string:
		return value.(string), nil
	default:
		return string(body), nil
	}
```

Inside SensorValue I can decide if it's a string, and return it as such if it is; otherwise I return the original (pre-Unmarshall) bytes.  If I am to do the check outside SensorValue then I will have to return the results of the Unmarshall.  Once that's checked by client code, it's too late to get at the original (pre-unmarshall) bytes, which means all the client has to go on is the Go data structure.  To represent this as JSON the client has to do its own explicit conversion.  I've put this into a utility, however, so it's not too awkward. 

Let me know what you think...

robertgmoss commented:
Yeah, I see what you mean, but despite the re-marshalling, I think this is better - it may be interesting to think about how else we could represent the non string data, rather than just converting it back to JSON - or finding a pretty printer - but we can do that in another PR.

I'm going to merge this now.
